### PR TITLE
tuftool updates for Delegated Targets

### LINKED ISSUE #5: Implement target delegation
The [TUF spec](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md) (section 4.5) defines a `delegations` object which lists paths and signing keys. tough should support this.

### PR DESCRIPTION
**Pull request by ecpullen**
**This is the last part of a set of 3 PRs to finish implementing delegated targets.**

*Issue #, if available:*
Closes #5 

*Description of changes:*
* Adds 6 tuftool subcommands related to delegated targets
* `tuftool delegations --signing-role <signing role> <subcommand>
Commands
* `create-role` creates a new `Targets` metadata file
* `add-role` adds a new `Targets` metadata file to the roles delegated roles
* `add-key` adds a key for signing the specified role
* `remove-key` removes a key used to sign the specified role
* `update-delegated-targets` adds targets and refreshes a `Targets` metadata
Any output from a delegations subcommand other than `create-role` must be added to the repository by a repo owner by using `update`
* `create-role` must be added to the delegating role by using `add-role` before the metadata can be sent to the repo owner.

*Testing*
Significant testing of the new commands can be found in `delegations_commands.rs`.
Tests include
* Creating a role
* Delegating a created role
* Updating a delegated targets role
* Adding a key to a delegated targets
* Removing a key from a delegated targets
* Removing an immediate delegated targets from a role
* Removing a delegated targets that is eventually delegated from a role

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tuftool/src/main.rs [hunk 1]
```diff
@@ -71,9 +77,11 @@ enum Command {
     /// Download a TUF repository's resources
     Download(download::DownloadArgs),
     /// Update a TUF repository's metadata and optionally add targets
-    Update(update::UpdateArgs),
+    Update(Box<update::UpdateArgs>),
```
Comment (zmrow) [inline tuftool/src/main.rs:80]: How come we had to Box here?

File: tuftool/src/main.rs [hunk 2]
```diff
@@ -163,3 +172,47 @@ fn main() -> ! {
...
+
+impl Delegation {
+    fn run(self) -> Result<()> {
+        self.cmd.run(&self.role)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+enum DelegationCommand {
+    CreateRole(Box<create_role::CreateRoleArgs>),
```
Comment (zmrow) [inline tuftool/src/main.rs:195]: Missing a docstring on this one. :) 

(even though it's fairly obvious let's keep it uniform.)

File: tuftool/src/main.rs [hunk 3]
```diff
@@ -163,3 +172,47 @@ fn main() -> ! {
...
+    /// Add delegated role
+    AddRole(Box<add_role::AddRoleArgs>),
+    /// Update Delegated targets
+    UpdateDelegatedTargets(Box<update_targets::UpdateTargetsArgs>),
+    /// Add a key to a delegatee
+    AddKey(Box<add_key_role::AddKeyArgs>),
+    /// Add a key to a delegatee
```
Comment (zmrow) [inline tuftool/src/main.rs]: I think this should be `Remove a key from a delegatee`

File: tuftool/src/add_role.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    #[structopt(short = "o", long = "outdir")]
+    outdir: PathBuf,
+
+    /// The delegated paths
+    #[structopt(short = "p", long = "paths")]
+    paths: Option<Vec<String>>,
+
+    /// The delegated paths hash prefixes
+    #[structopt(short = "hp", long = "path-hash-prefixes")]
+    path_hash_prefixes: Option<Vec<String>>,
```
Comment (zmrow) [inline tuftool/src/add_role.rs:67]: I think these are mutually exclusive right? If so, structopt gives you the capability to ensure only one of these arguments gets passed.

File: tuftool/src/add_role.rs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            .context(error::DelegationStructure)?
+            .targets_expires(self.expires)
+            .context(error::DelegationStructure)?
+            .sign_targets_editor(&self.keys)
+            .context(error::DelegateeNotFound {
+                role: role.to_string(),
+            })?
+            .change_delegated_targets(role)
+            .context(error::DelegateeNotFound {
+                role: role.to_string(),
```
Comment (zmrow) [inline tuftool/src/add_role.rs]: I know it's slightly longer, but could we split this call into 2 steps with comments to make it clear what's happening?

1) Update `Targets` version, expiration, sign
2) Change targets role to delegated targets

File: tuftool/src/error.rs [hunk 6]
```diff
@@ -60,12 +60,19 @@ pub(crate) enum Error {
...
     },
 
-    #[snafu(display("Invalid delegation structure"))]
+    #[snafu(display("Invalid delegation structure: {}", source))]
     DelegationStructure {
         source: tough::error::Error,
         backtrace: Backtrace,
     },
 
+    #[snafu(display("Couldn't find {}: {}", role, source))]
```
Comment (zmrow) [inline tuftool/src/error.rs]: Nit: `Couldn't fine role '{}': {}`

File: tuftool/src/remove_key_role.rs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+use tough::{ExpirationEnforcement, FilesystemTransport, Limits, Repository};
+use url::Url;
+
+#[derive(Debug, StructOpt)]
+pub(crate) struct RemoveKeyArgs {
+    /// Key files to sign with
+    #[structopt(short = "k", long = "key", required = true, parse(try_from_str = parse_key_source))]
+    keys: Vec<Box<dyn KeySource>>,
+
+    /// Keys to be removes
```
Comment (zmrow) [inline tuftool/src/remove_key_role.rs]: :%s/removes/removed

Also, could we add a hint for the user so they know what the argument should be?

File: tuftool/src/remove_key_role.rs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                role,
+                TargetsEditor::from_repo(&repository, role)
+                    .context(error::EditorFromRepo { path: &self.root })?,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    /// Adds keys to a role using targets Editor
```
Comment (zmrow) [inline tuftool/src/remove_key_role.rs]: I think this docstring is incorrect

File: tuftool/src/remove_role.rs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                role,
+                TargetsEditor::from_repo(&repository, role)
+                    .context(error::EditorFromRepo { path: &self.root })?,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    /// Adds keys to a role using targets Editor
```
Comment (zmrow) [inline tuftool/src/remove_role.rs]: This docstring is wrong. :)

File: tuftool/src/update.rs [hunk 10]
```diff
@@ -157,6 +165,20 @@ impl UpdateArgs {
...
         };
 
+        // If a `Targets` metadata needs to be updated
+        if self.role.is_some() && self.indir.is_some() {
+            editor
+                .sign_targets_editor(&self.keys)
+                .context(error::DelegationStructure)?
+                .update_delegated_targets(
+                    &self.role.as_ref().unwrap(),
+                    &self.indir.as_ref().unwrap().as_str(),
```
Comment (zmrow) [inline tuftool/src/update.rs]: Let's get rid of these `unwrap`s.

**Review by zmrow — COMMENTED** [review]
Nice work!

I didn't get all the way through the tests file, but here's what I've got so far. 👍

File: tuftool/src/main.rs (hunk 1 above)
Comment (ecpullen) [reply tuftool/src/main.rs:80]: There are much more arguments for update than other commands, so it was required to be put in a box to save space.

**Review by zmrow — APPROVED** [review]
Nice work! I'm good with this.

🎖️

File: tuftool/tests/delegation_commands.rs [hunk 11]
```diff
@@ -0,0 +1,1490 @@
...
+            "--timestamp-expires",
+            timestamp_expiration.to_rfc3339().as_str(),
+            "--timestamp-version",
+            format!("{}", timestamp_version).as_str(),
+        ])
+        .assert()
+        .success();
+}
+
+#[test]
```
Comment (webern) [inline tuftool/tests/delegation_commands.rs:56]: Nice job testing comprehensive use cases!

(I wish I could make a comment on an entire file.)

File: tuftool/src/main.rs (hunk 1 above)
Comment (webern) [reply tuftool/src/main.rs:80]: Clippy I suppose... since it's a union under the hood.

File: tuftool/src/main.rs [hunk 12]
```diff
@@ -163,3 +172,48 @@ fn main() -> ! {
...
+
+impl Delegation {
+    fn run(self) -> Result<()> {
+        self.cmd.run(&self.role)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+enum DelegationCommand {
+    // Creates a delegated role
```
Comment (webern) [inline tuftool/src/main.rs]:
```suggestion
    /// Creates a delegated role
```

File: tuftool/src/create.rs [hunk 13]
```diff
@@ -87,19 +88,23 @@ impl CreateArgs {
         }
 
         let targets = build_targets(&self.targets_indir, self.follow)?;
-        let mut editor =
-            RepositoryEditor::new(&self.root).context(error::EditorCreate { path: &self.root })?;
+        let mut editor = RepositoryEditor::<FilesystemTransport>::new(&self.root)
```
Comment (webern) [inline tuftool/src/create.rs:91]: I wonder what happened that caused clippy or rustc to stop inferring the type! Maybe someday clippy will make us change it all back again...

```rust
        let mut editor = RepositoryEditor::<FilesystemTransport>::new(&self.root)
```

File: tuftool/src/add_role.rs [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    snapshot_version: Option<NonZeroU64>,
+    /// Expiration of snapshot.json file; can be in full RFC 3339 format, or something like 'in
+    /// 7 days'
+    #[structopt(long = "snapshot-expires", parse(try_from_str = parse_datetime))]
+    snapshot_expires: Option<DateTime<Utc>>,
+
+    /// Version of timestamp.json file
+    #[structopt(long = "timestamp-version")]
+    timestamp_version: Option<NonZeroU64>,
+    /// Expiration of timestamp.json file; can be in full RFC 3339 format, or something like 'in
```
Comment (webern) [inline tuftool/src/add_role.rs]: nit for consistency
```suggestion
    timestamp_version: Option<NonZeroU64>,
    
    /// Expiration of timestamp.json file; can be in full RFC 3339 format, or something like 'in
```

File: tough/tests/repo_editor.rs [hunk 15]
```diff
@@ -1,16 +1,22 @@
 // Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
+use crate::test_utils::{dir_url, read_to_end, test_data};
 use chrono::{Duration, Utc};
+use std::collections::HashMap;
 use std::fs::File;
-use std::io::Read;
+use std::io::prelude::Write;
```
Comment (webern) [inline tough/tests/repo_editor.rs:8]:
```suggestion
use std::io::Write;
```

File: tough/src/editor/targets.rs [hunk 16]
```diff
@@ -0,0 +1,564 @@
...
+
+const SPEC_VERSION: &str = "1.0.0";
+
+/// If you are not working with a repository that utilizes delegated targets, use the `RepositoryEditor`.
+///
+/// `TargetsEditor` contains the various bits of data needed to construct
+/// or edit a `Targets` role.
+///
+/// `TargetsEditor` should only be used with repositories that utilize delegated targets.
+/// `TargetsEditor` cannot create a `SignedRepository`. Whenever a user has access to `snapshot.json` and `timestamp.json` keys,
```
Comment (webern) [inline tough/src/editor/targets.rs]: Nit, this doesn't seemed to be wrapped at the width of the rest of the documentation.

```rust
/// `TargetsEditor` cannot create a `SignedRepository`. Whenever a user has access to `snapshot.json` and `timestamp.json` keys,
```

File: tough/src/editor/targets.rs [hunk 17]
```diff
@@ -0,0 +1,564 @@
...
+/// `RepositoryEditor` should be used so that a `SignedRepository` can be created.
+///
+/// A new Targets may be created using the `new()` method.
+///
+/// An existing `Targets` may be loaded and edited using the
+/// `from_targets()` method. When a targets is loaded in this way, versions and
+/// expirations are discarded. It is good practice to update these whenever
+/// a repo is changed.
+///
+/// A  `Targets` from an existing repository can be loaded using the `from_repo()` method.
```
Comment (webern) [inline tough/src/editor/targets.rs]: nit
```suggestion
/// A `Targets` from an existing repository can be loaded using the `from_repo()` method.
```

**Review by webern — APPROVED** [review]
Nice work, Ethan.

**Comment by ecpullen** [comment]
Forced push rebases onto repoeditor pr

**Pull request merged**